### DISCUSSION TITLE
Why is matchSourceUrlPrefixes not working?

### DISCUSSION THREAD

**Discussion by timmkrause** [renovatebot/renovate#12374] [Request Help] [answered]
All `packageRules` of the below configuration file work perfectly, except the last one which is based on `matchSourceUrlPrefixes`.

Renovate version: 28.11.0

We're running on Azure DevOps Server 2020 and self-hosted agents.

Any ideas what could potentially cause this rule not to be applied?

**default.json**
```
{
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "assignAutomerge": true,
    "azureAutoComplete": true,
    "platformAutomerge": true,
    "autoMerge": true,
    "automergeType": "pr",
    "dependencyDashboard": false,
    "azureWorkItemId": 41037,
    "fetchReleaseNotes": false,
    "stabilityDays": 7,
    "prCreation": "not-pending",
    "internalChecksFilter": "strict",
    "packageRules": [
        {
            "descsription":"This can be updated when we switch to .NET 6",
            "matchPackageNames": ["Microsoft.NET.Sdk.Functions"],
            "allowedVersions": "<4.0.0"
        },
        {
            "description": "This can be updated when we switch to .NET 5",
            "matchPackagePrefixes": ["Microsoft", "System"],
            "allowedVersions": "<5.0.0"
        },
        {
            "description": "This can be updated when we switch to .NET 5, [...]",
            "matchPackageNames": ["Microsoft.Azure.Devices.Client"],
            "allowedVersions": "<1.38"
        },        
        {
            "description": "Group all Microsoft packages into single PRs to reduce noise",
            "matchPackagePrefixes": ["Microsoft", "System", "Azure", "dotnet-ef", "MSTest"],
            "groupName": "Microsoft dependencies"
        },
        {
            "description": "Group all Swashbuckle packages into single PRs to reduce noise",
            "matchPackagePrefixes": ["Swashbuckle"],
            "groupName": "Swashbuckle dependencies"
        },
        {
            "description": "Group all Google (mainly protobuf) packages into single PRs to reduce noise",
            "matchPackagePrefixes": ["Google", "Grpc"],
            "groupName": "Google dependencies"
        },
        {
            "description": "Group all COMPANY package into single PRs to reduce noise",
            "matchSourceUrlPrefixes": ["https://my.company.com"],
            "groupName": "COMPANY dependencies"
        }
    ]
}
```

**Comment by rarkins** [comment, marked as answer]
It's not really possible to guess why. Can you reproduce it in a public repository?

**Reply by viceice** [reply]
I guess renovate can't detect the sourceurl because the custom registry don't provide it.

I see he is using nuget, so if the registry is a v2, renovate can't detect sourceurl, so it's not possible to filter by.

**Reply by timmkrause** [reply]
Exactly, we're consuming NuGet packages via the Azure DevOps (Server) Artifacts Feeds.
It's a v3 endpoint.

How could it be, that it's not detectable? The updates itself are working. Renovate is able to grab newer versions and create PRs from it. If this is possible Renovate should also know from which source it came from, right?

**Reply by rarkins** [reply]
> If this is possible Renovate should also know from which source it came from, right?

Renovate doesn't need to know the source repository in order to do updates. So if the source repository URL is not available from the datasource, you won't get release notes. Do the PRs contain a link to the source repository?

**Reply by viceice** [reply]
this is the code to fetch `sourceUrl`
https://github.com/renovatebot/renovate/blob/9afdfdb09f91701b005430885b369a024aadc680/lib/datasource/nuget/v3.ts#L166-L181

So there is maybe a missing api on azure

**Reply by rarkins** [reply]
To match source url prefixes means that we must know the source URL. One way to determine if we know the source URL is if it's linked from the PR updates table.

**Reply by timmkrause** [reply]
Unfortunately I don't know how this works under the covers so I don't understand why the source URL is not known. At some point a request for packages and their versions must be made to detect versions which should (hopefully by any source) be confirmed by 2xx, my assumption would be that this is then the correct source URL. Or is this information not bubbled up and therefore not known? I don't think you're using native (in my world) NuGet/`dotnet` tooling which may hide this detail from Renovate?
Stupid question, just trying to follow you.

With .NET 6 now there is potentially also a static way determining the source via package source mapping (would require Renovate to know a little more about the `nuget.config`):

- https://www.youtube.com/watch?v=G6P38Dn69Ro
- https://devblogs.microsoft.com/nuget/introducing-package-source-mapping/

How is Renovates philosophy regarding these package manager specific approaches? Is this conceivable?

**Reply by viceice** [reply]
we fetch the `nuspec` file via nuget v3 api and read the `metadata.repository@url`  from it as `sourceUrl` if available

**Reply by viceice** [reply]
`sourceUrl` is not the registry url, where the package is hosted, it's the url to the source code of that package.

**Reply by timmkrause** [reply]
> 
> 
> `sourceUrl` is not the registry url, where the package is hosted, it's the url to the source code of that package.

Oh dear, that explains a lot. Thanks for pointing this out.
